src/main/java/build/buildfarm/worker/OutputDirectory.java

```diff
@@ -0,0 +1,122 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+public class OutputDirectory {
+  private final Map<String, OutputDirectory> children;
+
+  private static final OutputDirectory defaultInstance = new OutputDirectory(ImmutableMap.<String, OutputDirectory>of());
+  private static OutputDirectory getDefaultInstance() {
+    return defaultInstance;
+  }
+
+  private OutputDirectory(Map<String, OutputDirectory> children) {
+    this.children = children;
+  }
+
+  public static OutputDirectory parse(Iterable<String> outputFiles, Iterable<String> outputDirs) {
+    return parseDirectories(Iterables.concat(
+        ImmutableList.<Iterable<String>>of(
+            Iterables.transform(
+                Iterables.filter(outputFiles, (file) -> file.contains("/")),
+                (file) -> "/" + file.substring(0, file.lastIndexOf('/') + 1)),
+            Iterables.transform(outputDirs, (d) -> d.isEmpty() ? "/" : ("/" + d + "/")))));
+  }
+
+  private static class Builder {
+    Map<String, Builder> children = new HashMap<>();
+
+    public Builder addChild(String name) {
+      Builder childBuilder = new Builder();
+      Preconditions.checkState(!children.containsKey(name), "Duplicate Output Directory Name");
+      children.put(name, childBuilder);
+      return childBuilder;
+    }
+
+    public OutputDirectory build() {
+      if (children.isEmpty()) {
+        return OutputDirectory.getDefaultInstance();
+      }
+      return new OutputDirectory(ImmutableMap.copyOf(Maps.transformValues(children, (v) -> v.build())));
+    }
+  };
+
+  private static OutputDirectory parseDirectories(Iterable<String> outputDirs) {
+    Builder builder = new Builder();
+    Stack<Builder> stack = new Stack<>();
+
+    List<String> sortedOutputDirs = new ArrayList<>();
+    Iterables.addAll(sortedOutputDirs, outputDirs);
+    Collections.sort(sortedOutputDirs);
+
+    Builder currentBuilder = builder;
+    String prefix = "/";
+    for (String outputDir : sortedOutputDirs) {
+      while (!outputDir.startsWith(prefix)) {
+        currentBuilder = stack.pop();
+        int upPathSeparatorIndex = prefix.lastIndexOf('/', prefix.length() - 2);
+        prefix = prefix.substring(0, upPathSeparatorIndex + 1);
+      }
+      String prefixedFile = outputDir.substring(prefix.length(), outputDir.length() - 1);
```

Comment:
Code fails at line 87 here with a string out of bounds exception when the outputDir is identical to the prefix.  I worked around it with a `if (outputDir.length() == prefix.length()) { continue; }`.  WIth that additional change this PR appears to work for me.